In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LogisticRegression
from sklearn import linear_model
from sklearn import metrics
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler
from sklearn import preprocessing
from sklearn.cluster import KMeans
%matplotlib inline

# QUESTION 1: 

In this task, you have to work with a dataset in the file students.csv. The data provides
some information about students, including: (i) gender, (ii) the educational background of the student’s
parents, (iii) whether the student has participated in test preparation courses, and (iv) the test scores
in math, reading and writing. By making use of the basic data analysis tools introduced in the course,
answer the following questions:

In [3]:
# Import the dataset
student = pd.read_csv('students.csv')
student

gender  parental_education test_prep_course  math_score  reading_score  \
0    female   bachelor's degree             none          72             72   
1    female        some college        completed          69             90   
2    female     master's degree             none          90             95   
3      male  associate's degree             none          47             57   
4      male        some college             none          76             78   
..      ...                 ...              ...         ...            ...   
995  female     master's degree        completed          88             99   
996    male         high school             none          62             55   
997  female         high school        completed          59             71   
998  female        some college        completed          68             78   
999  female        some college             none          77             86   

     writing_score  
0               74  
1               88  
2               93  
3               44  
4               75  
..             ...  
995             95  
996             55  
997             65  
998             77  
999             86  

[1000 rows x 6 columns]

In [4]:
# Check the types of variables in the dataset
student.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   gender              1000 non-null   object
 1   parental_education  1000 non-null   object
 2   test_prep_course    1000 non-null   object
 3   math_score          1000 non-null   int64 
 4   reading_score       1000 non-null   int64 
 5   writing_score       1000 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 47.0+ KB


From the info of the data, we can see that there is no missing values

In [5]:
student.isnull()

gender  parental_education  test_prep_course  math_score  reading_score  \
0     False               False             False       False          False   
1     False               False             False       False          False   
2     False               False             False       False          False   
3     False               False             False       False          False   
4     False               False             False       False          False   
..      ...                 ...               ...         ...            ...   
995   False               False             False       False          False   
996   False               False             False       False          False   
997   False               False             False       False          False   
998   False               False             False       False          False   
999   False               False             False       False          False   

     writing_score  
0            False  
1            False  
2            False  
3            False  
4            False  
..             ...  
995          False  
996          False  
997          False  
998          False  
999          False  

[1000 rows x 6 columns]

Again, from the info of the data, we can see that NO missing values

In [6]:
student.isnull().sum()

gender                0
parental_education    0
test_prep_course      0
math_score            0
reading_score         0
writing_score         0
dtype: int64

In [7]:
student.head()

gender  parental_education test_prep_course  math_score  reading_score  \
0  female   bachelor's degree             none          72             72   
1  female        some college        completed          69             90   
2  female     master's degree             none          90             95   
3    male  associate's degree             none          47             57   
4    male        some college             none          76             78   

   writing_score  
0             74  
1             88  
2             93  
3             44  
4             75

In [8]:
student.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

• Are there more male or female students included in the dataset?

In [9]:
# From the info of the data, we can see that there is no missing values
student.corr()

math_score  reading_score  writing_score
math_score       1.000000       0.817580       0.802642
reading_score    0.817580       1.000000       0.954598
writing_score    0.802642       0.954598       1.000000

In [10]:
print('female students included in the dataset',student['gender'].tolist().count('female'))
print('male students included in the dataset',student['gender'].tolist().count('male'))

female students included in the dataset 518
male students included in the dataset 482


• In which subject did the students perform best overall (i.e. what is the subject with the highest
average score)?


In [49]:
student.groupby(['gender'])['math_score', 'reading_score', 'writing_score'].agg([np.max, np.min])

<ipython-input-49-ce6fd10efc1d>:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  student.groupby(['gender'])['math_score', 'reading_score', 'writing_score'].agg([np.max, np.min])


math_score      reading_score      writing_score     
             amax amin          amax amin          amax amin
gender                                                      
female        100    0           100   17           100   10
male          100   27           100   23           100   15

• How many students can you find with perfect results (100 score in all three subjects)?


In [12]:
student[(student['math_score'] > 99) & 
        (student['writing_score'] > 99) &
        (student['reading_score'] > 99)]

gender  parental_education test_prep_course  math_score  reading_score  \
458  female   bachelor's degree             none         100            100   
916    male   bachelor's degree        completed         100            100   
962  female  associate's degree             none         100            100   

     writing_score  
458            100  
916            100  
962            100

• Can we say that participating in the test preparation course improves the results in math (i.e. is the
average math score higher for those who completed the test preparation course)?

In [13]:
student[['test_prep_course',
         'gender',
         'math_score',
         'writing_score',
         'reading_score']].groupby(['test_prep_course','gender']).agg('median')

math_score  writing_score  reading_score
test_prep_course gender                                          
completed        female          67             79             78
                 male            73             70             71
none             female          62             70             71
                 male            67             60             63

# QUESTION 2:

In this exercise, you will have to analyze a dataset about some employees of a company
(salary.csv). The data includes information about (i) years of experience of the employee, (ii) the age of
the employee, and (iii) the salary of the employee. You need to write the code to answer the following
questions to transform and understand the data further.

a) There are some missing values in the data. In the experience years column, replace all missing values
by 3, and in the age column replace all the missing values by the mean of the column.


In [14]:
# Import the dataset
salary = pd.read_csv('salary.csv')
salary

experience_years   age  salary
0                1.1   NaN   39343
1                1.3  24.0   46205
2                1.5  28.7   37731
3                NaN  31.0   43525
4                2.2   NaN   39891
5                2.9  33.0   56642
6                3.0   NaN   60150
7                3.2  34.5   54445
8                NaN  28.3   64445
9                3.7  29.6   57189
10               3.9  28.9   63218
11               4.0  34.0   55794
12               4.0  29.5   56957
13               4.1  24.8   57081
14               4.5  35.0   61111
15               4.9  29.2   67938
16               NaN  26.0   66029
17               5.3   NaN   83088
18               5.9  28.0   81363
19               6.0  29.3   93940
20               6.8  33.2   91738
21               7.1  30.0   98273
22               NaN  31.4  101302
23               8.2  32.0  113812
24               8.7   NaN  109431
25               9.0  34.0  105582
26               NaN  35.0  116969
27               9.6  36.0  112635
28              10.3  37.0  122391
29              10.5  38.0  121872

In [15]:
# Check the types of variables in the dataset
# gender and edu are categorical values while others are numerics
salary.head()

experience_years   age  salary
0               1.1   NaN   39343
1               1.3  24.0   46205
2               1.5  28.7   37731
3               NaN  31.0   43525
4               2.2   NaN   39891

In [16]:
salary.shape

(30, 3)

In [17]:
salary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   experience_years  25 non-null     float64
 1   age               25 non-null     float64
 2   salary            30 non-null     int64  
dtypes: float64(2), int64(1)
memory usage: 848.0 bytes


In [18]:
salary.describe()

experience_years        age         salary
count         25.000000  25.000000      30.000000
mean           5.268000  31.216000   76003.000000
std            2.841613   3.712892   27414.429785
min            1.100000  24.000000   37731.000000
25%            3.200000  28.900000   56720.750000
50%            4.500000  31.000000   65237.000000
75%            7.100000  34.000000  100544.750000
max           10.500000  38.000000  122391.000000

In [19]:
experience_data = pd.Series(salary['experience_years'])

experience_data

0      1.1
1      1.3
2      1.5
3      NaN
4      2.2
5      2.9
6      3.0
7      3.2
8      NaN
9      3.7
10     3.9
11     4.0
12     4.0
13     4.1
14     4.5
15     4.9
16     NaN
17     5.3
18     5.9
19     6.0
20     6.8
21     7.1
22     NaN
23     8.2
24     8.7
25     9.0
26     NaN
27     9.6
28    10.3
29    10.5
Name: experience_years, dtype: float64

In [20]:
experience_data.isnull()

0     False
1     False
2     False
3      True
4     False
5     False
6     False
7     False
8      True
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16     True
17    False
18    False
19    False
20    False
21    False
22     True
23    False
24    False
25    False
26     True
27    False
28    False
29    False
Name: experience_years, dtype: bool

In [21]:
# If we take the sum of the above expression, we get the number of missing values

experience_data.isnull().sum()

5

As we can see it is 5

In [22]:
experience_data.fillna(3)

0      1.1
1      1.3
2      1.5
3      3.0
4      2.2
5      2.9
6      3.0
7      3.2
8      3.0
9      3.7
10     3.9
11     4.0
12     4.0
13     4.1
14     4.5
15     4.9
16     3.0
17     5.3
18     5.9
19     6.0
20     6.8
21     7.1
22     3.0
23     8.2
24     8.7
25     9.0
26     3.0
27     9.6
28    10.3
29    10.5
Name: experience_years, dtype: float64

----in the age column replace all the missing values by the mean of the column

In [23]:
age_data = pd.Series(salary['age'])

age_data

0      NaN
1     24.0
2     28.7
3     31.0
4      NaN
5     33.0
6      NaN
7     34.5
8     28.3
9     29.6
10    28.9
11    34.0
12    29.5
13    24.8
14    35.0
15    29.2
16    26.0
17     NaN
18    28.0
19    29.3
20    33.2
21    30.0
22    31.4
23    32.0
24     NaN
25    34.0
26    35.0
27    36.0
28    37.0
29    38.0
Name: age, dtype: float64

In [24]:
age_data.fillna(age_data.mean())

0     31.216
1     24.000
2     28.700
3     31.000
4     31.216
5     33.000
6     31.216
7     34.500
8     28.300
9     29.600
10    28.900
11    34.000
12    29.500
13    24.800
14    35.000
15    29.200
16    26.000
17    31.216
18    28.000
19    29.300
20    33.200
21    30.000
22    31.400
23    32.000
24    31.216
25    34.000
26    35.000
27    36.000
28    37.000
29    38.000
Name: age, dtype: float64

In [25]:
salary['age'].fillna(int(salary['age'].mean()), inplace=True)
salary['experience_years'].fillna(3, inplace=True)

In [26]:
print(salary)

    experience_years   age  salary
0                1.1  31.0   39343
1                1.3  24.0   46205
2                1.5  28.7   37731
3                3.0  31.0   43525
4                2.2  31.0   39891
5                2.9  33.0   56642
6                3.0  31.0   60150
7                3.2  34.5   54445
8                3.0  28.3   64445
9                3.7  29.6   57189
10               3.9  28.9   63218
11               4.0  34.0   55794
12               4.0  29.5   56957
13               4.1  24.8   57081
14               4.5  35.0   61111
15               4.9  29.2   67938
16               3.0  26.0   66029
17               5.3  31.0   83088
18               5.9  28.0   81363
19               6.0  29.3   93940
20               6.8  33.2   91738
21               7.1  30.0   98273
22               3.0  31.4  101302
23               8.2  32.0  113812
24               8.7  31.0  109431
25               9.0  34.0  105582
26               3.0  35.0  116969
27               9.6

b) Check the correlation between the columns of the data. Based on the results, what is your opinion,
does experience or age impact salary more?

   - Check the correlation between the columns of the data

In [27]:
salary.corr()

experience_years       age    salary
experience_years          1.000000  0.535703  0.834603
age                       0.535703  1.000000  0.523353
salary                    0.834603  0.523353  1.000000

In [28]:
#Calculate the correlation between Experience years	and Age
exp_age=salary['experience_years'].corr(salary['age'])
print('correlation between Experience years and Age is', exp_age)

correlation between Experience years and Age is 0.5357028471981783


In [29]:
#Calculate the correlation between Experience years	and Salary
exp_salary=salary['experience_years'].corr(salary['salary'])
print('correlation between Experience years and Salary is', exp_salary)

correlation between Experience years and Salary is 0.8346030993050744


In [30]:
#Calculate the correlation between Salary years	and Age
salary_age=salary['salary'].corr(salary['age'])
print('correlation between Salary and Age is', salary_age)

correlation between Salary and Age is 0.523352990905591


???Based on the results, what is your opinion, does experience or age impact salary more?

In [31]:
np.corrcoef(salary['experience_years'], salary['age'])

array([[1.        , 0.53570285],
       [0.53570285, 1.        ]])

 --->>>Correlation coefficients whose magnitude are between 0.7 and 0.9 indicate variables which can be considered highly correlated. Correlation coefficients whose magnitude are between 0.5 and 0.7 indicate variables which can be considered moderately correlated

In [32]:
from scipy.stats import linregress
linregress(salary['experience_years'], salary['age'])

LinregressResult(slope=0.6643869031301838, intercept=27.9311480436934, rvalue=0.5357028471981783, pvalue=0.002281606945714231, stderr=0.1979108925555436)

Since the p-value of 0.0022 is smaller than 0.05, we reject the null hypothesis that the relationship between the age and their work experience is not significant.

- Create a categorical version of the salary column with 3 groups: (i) values between 0 and 55000, (ii)
values between 55000 and 90000, (iii) values between 90000 and 125000. Label the groups [’Low’,
’Medium’, ’High’]. Calculate the average age of employees who have High salary.

In [33]:
Low=salary[(salary['salary']>0) & (salary['salary']<55000)]
Medium=salary[(salary['salary']>55000) & (salary['salary']<90000)]
High=salary[(salary['salary']>90000) & (salary['salary']<125000)]
print ('Group LOW', 
       Low)
print ('Group MEDIUM', 
       Medium)
print ('Group HIGH', 
       High)

Group LOW    experience_years   age  salary
0               1.1  31.0   39343
1               1.3  24.0   46205
2               1.5  28.7   37731
3               3.0  31.0   43525
4               2.2  31.0   39891
7               3.2  34.5   54445
Group MEDIUM     experience_years   age  salary
5                2.9  33.0   56642
6                3.0  31.0   60150
8                3.0  28.3   64445
9                3.7  29.6   57189
10               3.9  28.9   63218
11               4.0  34.0   55794
12               4.0  29.5   56957
13               4.1  24.8   57081
14               4.5  35.0   61111
15               4.9  29.2   67938
16               3.0  26.0   66029
17               5.3  31.0   83088
18               5.9  28.0   81363
Group HIGH     experience_years   age  salary
19               6.0  29.3   93940
20               6.8  33.2   91738
21               7.1  30.0   98273
22               3.0  31.4  101302
23               8.2  32.0  113812
24               8.7  31.0  

- Calculate the average age of employees who have High salary

In [34]:
print('The average age of employees who have High salary:')
High["age"].mean()

The average age of employees who have High salary:


33.35454545454545

# QUESTION 3

In this exercise, you will have to work with a dataset about movies (movies.csv). In the
dataset, you can find information about: (i) movie title, (ii) short overview of the movie, (iii) the original
language, (iv) number of votes given by people who have seen the movie, and (v) the average of the votes
for the movie (a vote is between 0 and 10). You need to write the code to answer the following questions.

In [121]:
movies = pd.read_csv('movies.csv')
movies

Unnamed: 0                           title  \
0              0                        Ad Astra   
1              1                       Bloodshot   
2              2               Bad Boys for Life   
3              3                         Ant-Man   
4              4  Percy Jackson: Sea of Monsters   
...          ...                             ...   
9995        9995                           Cargo   
9996        9996                  The Good Night   
9997        9997              The World Is Yours   
9998        9998             The Grand Seduction   
9999        9999        Woochi: The Demon Slayer   

                                               overview original_language  \
0     The near future, a time when both hope and har...                en   
1     After he and his wife are murdered, marine Ray...                en   
2     Marcus and Mike are forced to confront new thr...                en   
3     Armed with the astonishing ability to shrink i...                en   
4     In their quest to confront the ultimate evil, ...                en   
...                                                 ...               ...   
9995  The story of CARGO takes place on rusty space-...                de   
9996  Gary, a musician, is trapped in an unhappy rel...                en   
9997  To escape his life of crime, a Paris drug deal...                fr   
9998  A small fishing village must procure a local d...                en   
9999  Spanning four centuries in Korea, this epic ac...                ko   

      vote_count  vote_average  
0           2853           5.9  
1           1349           7.2  
2           2530           7.1  
3          13611           7.1  
4           3542           5.9  
...          ...           ...  
9995         225           5.9  
9996          67           5.6  
9997         234           7.1  
9998         169           6.7  
9999          78           6.7  

[10000 rows x 6 columns]

In [122]:
# We have header in the file, and we can use the first column, the Un, as index
movies.drop(['Unnamed: 0'],axis=1,inplace=True)


movies.head()

title  \
0                        Ad Astra   
1                       Bloodshot   
2               Bad Boys for Life   
3                         Ant-Man   
4  Percy Jackson: Sea of Monsters   

                                            overview original_language  \
0  The near future, a time when both hope and har...                en   
1  After he and his wife are murdered, marine Ray...                en   
2  Marcus and Mike are forced to confront new thr...                en   
3  Armed with the astonishing ability to shrink i...                en   
4  In their quest to confront the ultimate evil, ...                en   

   vote_count  vote_average  
0        2853           5.9  
1        1349           7.2  
2        2530           7.1  
3       13611           7.1  
4        3542           5.9

In [106]:
# When we look as the shape of the data, we can see that it has 10000 rows (countries) and 6 columns (attributes of countries)

movies.shape

(10000, 5)

In [107]:
# In order to look at the data, for the numerical columns, we can use the describe method
# The count tells us the number of available values, and as we can see, there are several columns with many missing values

movies.describe()

vote_count  vote_average
count  10000.000000  10000.000000
mean    1020.825100      6.306230
std     1992.305005      1.354259
min        0.000000      0.000000
25%      143.000000      5.800000
50%      332.000000      6.500000
75%      926.250000      7.100000
max    25148.000000     10.000000

In [108]:
movies.info ()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   title              10000 non-null  object 
 1   overview           9970 non-null   object 
 2   original_language  10000 non-null  object 
 3   vote_count         10000 non-null  int64  
 4   vote_average       10000 non-null  float64
dtypes: float64(1), int64(1), object(3)
memory usage: 390.8+ KB


In [109]:
movies.isnull()

title  overview  original_language  vote_count  vote_average
0     False     False              False       False         False
1     False     False              False       False         False
2     False     False              False       False         False
3     False     False              False       False         False
4     False     False              False       False         False
...     ...       ...                ...         ...           ...
9995  False     False              False       False         False
9996  False     False              False       False         False
9997  False     False              False       False         False
9998  False     False              False       False         False
9999  False     False              False       False         False

[10000 rows x 5 columns]

In [110]:
movies.isnull().sum()

title                 0
overview             30
original_language     0
vote_count            0
vote_average          0
dtype: int64

-Filter our missing data

In [111]:
# The basic command is dropna, with this we can delete all the missing values
# In our series, there will be 9970 values left

movies_full = movies.dropna()

len(movies_full)

9970

In [112]:
# As we can check, it has no missing values anymore

movies_full.isnull().sum()

title                0
overview             0
original_language    0
vote_count           0
vote_average         0
dtype: int64

In [113]:
# Alternatively, we can use the following code to drop all the missing values

movies[movies.notnull()]

title  \
0                           Ad Astra   
1                          Bloodshot   
2                  Bad Boys for Life   
3                            Ant-Man   
4     Percy Jackson: Sea of Monsters   
...                              ...   
9995                           Cargo   
9996                  The Good Night   
9997              The World Is Yours   
9998             The Grand Seduction   
9999        Woochi: The Demon Slayer   

                                               overview original_language  \
0     The near future, a time when both hope and har...                en   
1     After he and his wife are murdered, marine Ray...                en   
2     Marcus and Mike are forced to confront new thr...                en   
3     Armed with the astonishing ability to shrink i...                en   
4     In their quest to confront the ultimate evil, ...                en   
...                                                 ...               ...   
9995  The story of CARGO takes place on rusty space-...                de   
9996  Gary, a musician, is trapped in an unhappy rel...                en   
9997  To escape his life of crime, a Paris drug deal...                fr   
9998  A small fishing village must procure a local d...                en   
9999  Spanning four centuries in Korea, this epic ac...                ko   

      vote_count  vote_average  
0           2853           5.9  
1           1349           7.2  
2           2530           7.1  
3          13611           7.1  
4           3542           5.9  
...          ...           ...  
9995         225           5.9  
9996          67           5.6  
9997         234           7.1  
9998         169           6.7  
9999          78           6.7  

[10000 rows x 5 columns]

a) Evaluations that are based on too few votes cannot be completely trusted. For this reason, start
by removing movies that has less than 100 votes (vote count). Perform the following tasks on this
filtered dataset.

 - Filter movies that has less than 100 votes (vote count).

In [128]:
rating_by_users=movies.groupby(['title'])['vote_count'].size()
rating_by_users= pd.DataFrame(rating_by_users)
rating_by_users.head()

vote_count
title                           
#Iamhere                       1
#RealityHigh                   1
'71                            1
'83                            1
(500) Days of Summer           1

In [129]:
active_user= movies.loc[movies.vote_count>=100]
active_user

title  \
0                           Ad Astra   
1                          Bloodshot   
2                  Bad Boys for Life   
3                            Ant-Man   
4     Percy Jackson: Sea of Monsters   
...                              ...   
9992                  Eagle vs Shark   
9993                High Flying Bird   
9995                           Cargo   
9997              The World Is Yours   
9998             The Grand Seduction   

                                               overview original_language  \
0     The near future, a time when both hope and har...                en   
1     After he and his wife are murdered, marine Ray...                en   
2     Marcus and Mike are forced to confront new thr...                en   
3     Armed with the astonishing ability to shrink i...                en   
4     In their quest to confront the ultimate evil, ...                en   
...                                                 ...               ...   
9992  The New Zealand made "Eagle vs Shark" is the t...                en   
9993  During an NBA lockout, a sports agent, Ray Bur...                en   
9995  The story of CARGO takes place on rusty space-...                de   
9997  To escape his life of crime, a Paris drug deal...                fr   
9998  A small fishing village must procure a local d...                en   

      vote_count  vote_average  
0           2853           5.9  
1           1349           7.2  
2           2530           7.1  
3          13611           7.1  
4           3542           5.9  
...          ...           ...  
9992         154           6.6  
9993         124           5.8  
9995         225           5.9  
9997         234           7.1  
9998         169           6.7  

[8405 rows x 5 columns]

In [181]:
rating_user=active_user.groupby('vote_count').size()
rating_user

vote_count
100      30
101      24
102      31
103      23
104      20
         ..
21561     1
21589     1
21893     1
22225     1
25148     1
Length: 2673, dtype: int64

b) What is the 3rd most frequent original language in the data?



In [139]:
active_user["original_language"].value_counts()

en    7197
fr     328
ja     190
it     132
es     109
de      74
cn      54
hi      51
ko      50
zh      42
ru      34
da      27
sv      25
pt      18
no       9
tr       8
pl       7
nl       6
id       5
fa       5
th       4
sr       3
hu       3
he       2
el       2
ta       2
ar       2
te       2
fi       2
cs       2
uk       1
et       1
nb       1
la       1
ro       1
bs       1
eu       1
af       1
xx       1
sh       1
Name: original_language, dtype: int64

c) How many Finnish movies are in the data (original language is fi), and which is the highest rated
one?

In [182]:
active_user.original_language.unique()

array(['en', 'fr', 'es', 'tr', 'ko', 'it', 'ja', 'pl', 'cn', 'da', 'hi',
       'sv', 'el', 'de', 'ar', 'ru', 'zh', 'pt', 'no', 'th', 'sr', 'id',
       'te', 'xx', 'hu', 'sh', 'fa', 'la', 'nl', 'he', 'nb', 'bs', 'cs',
       'eu', 'fi', 'af', 'ta', 'et', 'uk', 'ro'], dtype=object)

In [185]:
finnish_movies=active_user.loc[movies.original_language.isin(['fi'])]
finnish_movies

title  \
6553          The Man Without a Past   
7142  Rare Exports: A Christmas Tale   

                                               overview original_language  \
6553  The second part of Aki Kaurismäki's "Finland" ...                fi   
7142  It's the eve of Christmas in northern Finland ...                fi   

      vote_count  vote_average  
6553         143           7.4  
7142         269           6.4

In [186]:
highest_rated=finnish_movies.pivot_table('vote_count', index='title', 
                          columns=('original_language'), aggfunc='size')
highest_rated

original_language               fi
title                             
Rare Exports: A Christmas Tale   1
The Man Without a Past           1

In [187]:
# the highest rated Finnish movies, from highest to lowesr
# based on vote_count
finnish_movies.sort_values(by='vote_count', ascending=False)[:3]

title  \
7142  Rare Exports: A Christmas Tale   
6553          The Man Without a Past   

                                               overview original_language  \
7142  It's the eve of Christmas in northern Finland ...                fi   
6553  The second part of Aki Kaurismäki's "Finland" ...                fi   

      vote_count  vote_average  
7142         269           6.4  
6553         143           7.4

d) Calculate the average of votes separately for all French (original language is fr) and Spanish (original
language is es) movies. Which country has higher average value?



In [188]:
active_user[["title", "original_language"]].describe()


title original_language
count                   8405              8405
unique                  8155                40
top     Beauty and the Beast                en
freq                       4              7197

Filter French and Spanish movies from orginal data

In [189]:
fr_es=active_user.loc[active_user.original_language.isin(['fr', 'es'])]
fr_es

title  \
7                                       Cold Blood   
9                                     The Platform   
137                                         Desire   
204                                   The Occupant   
225                     Portrait of a Lady on Fire   
...                                            ...   
9881                        The Phantom of Liberty   
9899  Belle and Sebastian: The Adventure Continues   
9942                                        Romina   
9967                           Belle and Sebastian   
9997                            The World Is Yours   

                                               overview original_language  \
7     A legendary but retired hit man lives in peace...                fr   
9     A mysterious place, an indescribable prison, a...                es   
137   In a social context deteriorated by a countryw...                fr   
204   An unemployed executive is forced to sell his ...                es   
225   On an isolated island in Brittany at the end o...                fr   
...                                                 ...               ...   
9881  This Surrealist film, with a title referencing...                fr   
9899  September, 1945. Sebastian impatiently waits f...                fr   
9942  A shocking assault inspires horrific violence ...                es   
9967  Belle and Sebastian is set high in the snowy A...                fr   
9997  To escape his life of crime, a Paris drug deal...                fr   

      vote_count  vote_average  
7            119           5.1  
9           1924           7.2  
137          215           4.7  
204          232           6.3  
225          477           8.2  
...          ...           ...  
9881         116           7.6  
9899         218           7.0  
9942         109           2.4  
9967         410           6.9  
9997         234           7.1  

[437 rows x 5 columns]

In [190]:
fr_es[["original_language", "vote_average"]].groupby("original_language").mean()


vote_average
original_language              
es                     6.823853
fr                     6.844207

From the above resilt, the average of votes of French movies is higher than Espaanjan ones

e) Create a categorical version of the vote average column with 4 groups: (i) values between 0 and 2.5,
(ii) values between 2.5 and 5, (iii) values between 5 and 7.5, and (iv) values between 7.5 and 10.
Label the groups [’Bad’, ’OK’, ’Good’, ’Excellent’]. Which of the four created groups has the highest
number of movies

In [195]:
new_data=active_user.groupby('title')['vote_average'].size()
new_data

title
#RealityHigh                  1
'71                           1
(500) Days of Summer          1
*batteries not included       1
...And Justice for All        1
                             ..
xXx                           1
xXx: Return of Xander Cage    1
xXx: State of the Union       1
¡Three Amigos!                1
Æon Flux                      1
Name: vote_average, Length: 8155, dtype: int64

-Create a categorical version of the vote average column with 4 groups

In [198]:
Bad=active_user[(active_user['vote_average']>0) & (active_user['vote_average']<2.5)]
OK=active_user[(active_user['vote_average']>2.5) & (active_user['vote_average']<5)]
Good=active_user[(active_user['vote_average']>5) & (active_user['vote_average']<7.5)]
Excellent=active_user[(active_user['vote_average']>7.5) & (active_user['vote_average']<10)]

In [199]:
print ('Group Bad', Bad)

Group Bad                          title  \
8751  Manos: The Hands of Fate   
9942                    Romina   

                                               overview original_language  \
8751  A family gets lost on the road and stumbles up...                en   
9942  A shocking assault inspires horrific violence ...                es   

      vote_count  vote_average  
8751         117           1.8  
9942         109           2.4  


In [200]:
print ('Group OK', OK)

Group OK                                      title  \
35             30 Days of Night: Dark Days   
108              Anacondas: Trail of Blood   
137                                 Desire   
345                                   Cats   
499   Sexual Chronicles of a French Family   
...                                    ...   
9908                       Accidental Love   
9917                             The Order   
9918                           The Clapper   
9920         Ace Ventura Jr: Pet Detective   
9985                        The Diabolical   

                                               overview original_language  \
35    After surviving the incidents in Barrow, Alask...                en   
108   A genetically created Anaconda, cut in half, r...                en   
137   In a social context deteriorated by a countryw...                fr   
345   A tribe of cats called the Jellicles must deci...                en   
499   Three generations of a French family open up a.

In [201]:
print ('Group Good', Good)

Group Good                                title  \
0                           Ad Astra   
1                          Bloodshot   
2                  Bad Boys for Life   
3                            Ant-Man   
4     Percy Jackson: Sea of Monsters   
...                              ...   
9992                  Eagle vs Shark   
9993                High Flying Bird   
9995                           Cargo   
9997              The World Is Yours   
9998             The Grand Seduction   

                                               overview original_language  \
0     The near future, a time when both hope and har...                en   
1     After he and his wife are murdered, marine Ray...                en   
2     Marcus and Mike are forced to confront new thr...                en   
3     Armed with the astonishing ability to shrink i...                en   
4     In their quest to confront the ultimate evil, ...                en   
...                                           

In [202]:
print ('Group Excellent', Excellent)

Group Excellent                                              title  \
14                                          Onward   
21                                           Joker   
22                           Miracle in Cell No. 7   
25                                        Parasite   
30    Harry Potter and the Deathly Hallows: Part 2   
...                                            ...   
9898                                 Pandora's Box   
9941                               The Diamond Arm   
9950                       The Ghost and Mrs. Muir   
9956                               Never Look Away   
9986                           Cool Kids Don't Cry   

                                               overview original_language  \
14    In a suburban fantasy world, two teenage elf b...                en   
21    During the 1980s, a failed stand-up comedian i...                en   
22    Separated from his daughter, a father with an ...                tr   
25    All unemployed, Ki-ta

In [209]:
Excellent.sum()
Bad.sum()
Good.sum()

title                Ad AstraBloodshotBad Boys for LifeAnt-ManPercy...
original_language    enenenenenenenfrenesenenenenenenenenenenenenen...
vote_count                                                     7125390
vote_average                                                   43182.5
dtype: object

In [231]:
movies_bins = [0, 2.5, 5, 7.5, 10]

In [232]:
# In our movies data, for example we could make the vote_average column categorical using the folowing bin limits


movies_bins = [0, 2.5, 5, 7.5, 10]

movies_cat = pd.cut(active_user.vote_average, movies_bins)

movies_cat

0       (5.0, 7.5]
1       (5.0, 7.5]
2       (5.0, 7.5]
3       (5.0, 7.5]
4       (5.0, 7.5]
           ...    
9992    (5.0, 7.5]
9993    (5.0, 7.5]
9995    (5.0, 7.5]
9997    (5.0, 7.5]
9998    (5.0, 7.5]
Name: vote_average, Length: 8405, dtype: category
Categories (4, interval[float64]): [(0.0, 2.5] < (2.5, 5.0] < (5.0, 7.5] < (7.5, 10.0]]

In [233]:
# We can check the frequency of categories with value_counts
movies_cat.value_counts()

(5.0, 7.5]     7008
(7.5, 10.0]     974
(2.5, 5.0]      421
(0.0, 2.5]        2
Name: vote_average, dtype: int64

Result: From the output, we can see that Group Good (5-7.5) having highest number of movies which are 7008 movies